libfwupdplugin/fu-backend.c

```diff
@@ -123,6 +123,35 @@ fu_backend_registered(FuBackend *self, FuDevice *device)
 		klass->registered(self, device);
 }
 
+/**
+ * fu_backend_invalidate:
+ * @self: a #FuBackend
+ *
+ * Normally when calling fu_backend_setup() multiple times it is only actually done once.
+ * Calling this method causes the next requests to fu_backend_setup() to actually probe the
+ * hardware.
+ *
+ * This should be done in case the backing information source has changed, for instance if
+ * a platform subsystem has been updated.
+ *
+ * This also optionally calls the ->invalidate() vfunc for the backend. This allows the backend
+ * to perform shared backend actions on superclassed devices.
+ *
+ * Since: 1.8.0
+ **/
+void
+fu_backend_invalidate(FuBackend *self)
+{
+	FuBackendClass *klass = FU_BACKEND_GET_CLASS(self);
+	FuBackendPrivate *priv = GET_PRIVATE(self);
+
+	g_return_if_fail(FU_IS_BACKEND(self));
+
+	priv->done_setup = FALSE;
+	if (klass->invalidate != NULL)
+		klass->invalidate(self);
```

Comment:
I think you probably want to prohibit certain backend types from this.  Like the built in ones shouldn't be doing this.

Reply:
> the built in ones shouldn't be doing this

Why not?

Reply:
What happens when a different plugin relies on it? If one USB plugin invalidates the USB backend and it takes down  the other USB devices too.

Reply:
Hmm, fair point. What about if fu_backend_device_added() saw more than one kind of plugin it would prevent invalidate (and set a GError?). Or maybe we just want a fu_backend_add_flag(backend, FU_BACKEND_FLAG_CAN_INVALIDATE) or something? Ideas welcome.

Reply:
What's the use case? I don't think I follow the background.

Reply:
> Or maybe we just want a fu_backend_add_flag(backend, FU_BACKEND_FLAG_CAN_INVALIDATE)

I think that's a reasonable solution.  Then if we find a need to modify other backends can think them through very closely.

Reply:
I'll just move it to the redfish plugin for now; we can do the complicated generic thing if any other backends need this functionality...

Reply:
Gah, I need the priv->done_setup = FALSE clearing, which needs to be in fu-backend.c. Back to plan A.

Reply:
Okay, there's now a construction-only property backends have to set for safety.